### PR TITLE
GRAD2-2853: initial implementation of student course CRUD is committed.

### PR DESCRIPTION
**Pull request by infstar**
GRAD2-2853: initial implementation of student course CRUD is committed.

### CODE REVIEW

File: api/src/main/java/ca/bc/gov/educ/api/course/controller/v2/CourseController.java [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+
+@CrossOrigin
+@RestController("CourseControllerV2")
+@Slf4j
+@RequestMapping(EducCourseApiConstants.GRAD_COURSE_URL_MAPPING_V2)
+@OpenAPIDefinition(info = @Info(title = "API for Student Course Data.", description = "This API is for Reading Student Course data.", version = "2"),
+        security = {@SecurityRequirement(name = "OAUTH2", scopes = {"READ_GRAD_STUDENT_COURSE_DATA"})})
+public class CourseController {
+
+    private static final String BEARER = "Bearer ";
```
Comment (cditcher) [inline api/src/main/java/ca/bc/gov/educ/api/course/controller/v2/CourseController.java:30]: Don't need BEARER now

File: api/src/main/java/ca/bc/gov/educ/api/course/controller/v2/StudentCourseController.java [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+
+@CrossOrigin
+@RestController("StudentCourseControllerV2")
+@Slf4j
+@RequestMapping(EducCourseApiConstants.STUDENT_COURSE_URL_MAPPING_V2)
+@OpenAPIDefinition(info = @Info(title = "API for Student Course Data.", description = "This API is for Reading Student Course data.", version = "2"),
+        security = {@SecurityRequirement(name = "OAUTH2", scopes = {"READ_GRAD_STUDENT_COURSE_DATA"})})
+public class StudentCourseController {
+
+    private static final String BEARER = "Bearer ";
```
Comment (cditcher) [inline api/src/main/java/ca/bc/gov/educ/api/course/controller/v2/StudentCourseController.java:32]: Remove BEARER

**Review by cditcher — CHANGES_REQUESTED** [review]

File: api/src/main/java/ca/bc/gov/educ/api/course/controller/v2/StudentCourseController.java [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+        if (studentCourseList.isEmpty()) {
+            return response.NO_CONTENT();
+        }
+        return response.GET(studentCourseList);
+    }
+
+    @PostMapping
+    @PreAuthorize(PermissionsConstants.UPDATE_GRAD_COURSE_RESTRICTION)
+    @Operation(summary = "Save a Student Course", description = "Save a Student Course", tags = { "Student Courses" })
+    @ApiResponses(value = {@ApiResponse(responseCode = "200", description = "OK"),
```
[the remark this answers is not in the document]
Comment (infstar) [reply api/src/main/java/ca/bc/gov/educ/api/course/controller/v2/StudentCourseController.java:69]: This one is also updating as well if it already exists.

File: api/src/main/java/ca/bc/gov/educ/api/course/controller/v2/StudentCourseController.java [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+    @GetMapping(EducCourseApiConstants.GET_STUDENT_COURSES_BY_STUDENT_ID_MAPPING)
+    @PreAuthorize(PermissionsConstants.READ_GRAD_COURSE)
+    @Operation(summary = "Find All Student Courses by Student ID", description = "Get All Student Courses by Student ID", tags = {"Student Courses"})
+    @ApiResponses(value = {@ApiResponse(responseCode = "200", description = "OK"), @ApiResponse(responseCode = "204", description = "NO CONTENT")})
+    public ResponseEntity<List<StudentCourse>> getStudentCoursesByStudentID(
+            @PathVariable UUID studentID, @RequestParam(value = "sortForUI", required = false, defaultValue = "false") boolean sortForUI, @RequestHeader(name="Authorization") String accessToken) {
+        validation.requiredField(studentID, "Student ID");
+        if (validation.hasErrors()) {
+            validation.stopOnErrors();
+            return new ResponseEntity<>(HttpStatus.BAD_REQUEST);
```
[the remark this answers is not in the document]
Comment (infstar) [reply api/src/main/java/ca/bc/gov/educ/api/course/controller/v2/StudentCourseController.java:56]: Inside "stopOnErrors()",  it throws GradBusinessRuleException, so the line 56 seems not reachable.

File: api/src/main/java/ca/bc/gov/educ/api/course/controller/v2/StudentCourseController.java [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+        }
+        log.debug("#Get All Student Course by Student ID: {}", studentID);
+        List<StudentCourse> studentCourseList = studentCourseService.getStudentCourses(studentID, sortForUI, accessToken.replace(BEARER, ""));
+        if (studentCourseList.isEmpty()) {
+            return response.NO_CONTENT();
+        }
+        return response.GET(studentCourseList);
+    }
```
[the remark this answers is not in the document]
Comment (infstar) [reply api/src/main/java/ca/bc/gov/educ/api/course/controller/v2/StudentCourseController.java:64]: saveStudentCourse method is for create & update.

File: api/src/main/java/ca/bc/gov/educ/api/course/controller/v2/StudentCourseController.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (infstar) [reply api/src/main/java/ca/bc/gov/educ/api/course/controller/v2/StudentCourseController.java:56]: which means validation error returns with 422.

File: api/src/main/java/ca/bc/gov/educ/api/course/controller/v2/CourseController.java [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+    public CourseController(@Qualifier("CourseServiceV2") CourseService courseService, GradValidation validation, ResponseHelper response) {
+        this.courseService = courseService;
+        this.validation = validation;
+        this.response = response;
+    }
+    @GetMapping(EducCourseApiConstants.GET_COURSE_BY_COURSE_ID_MAPPING)
+    @PreAuthorize(PermissionsConstants.READ_GRAD_COURSE)
+    @Operation(summary = "Find a Course by Course ID",
+            description = "Get a Course by Course ID", tags = { "Courses" })
+    @ApiResponses(value = {@ApiResponse(responseCode = "200", description = "OK")})
```
[the remark this answers is not in the document]
Comment (infstar) [reply api/src/main/java/ca/bc/gov/educ/api/course/controller/v2/CourseController.java:48]: done

File: api/src/main/java/ca/bc/gov/educ/api/course/controller/v2/CourseController.java [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+    public ResponseEntity<Course> getCourseDetails(@PathVariable String courseID, @RequestHeader(name="Authorization") String accessToken) {
+        log.info("#getCourseDetails : courseID={}", courseID);
+        return response.GET(courseService.getCourseInfo(courseID, accessToken.replace(BEARER, "")));
+    }
+
+    @GetMapping(EducCourseApiConstants.GET_COURSE_BY_CODE_MAPPING)
+    @PreAuthorize(PermissionsConstants.READ_GRAD_COURSE)
+    @Operation(summary = "Find a Course by Course Code and Course Level",
+            description = "Get a Course by Course Code and Course Level", tags = { "Courses" })
+    @ApiResponses(value = {@ApiResponse(responseCode = "200", description = "OK")})
```
[the remark this answers is not in the document]
Comment (infstar) [reply api/src/main/java/ca/bc/gov/educ/api/course/controller/v2/CourseController.java:58]: done

File: api/src/main/java/ca/bc/gov/educ/api/course/controller/v2/StudentCourseController.java (hunk 2 above)
Comment (infstar) [reply api/src/main/java/ca/bc/gov/educ/api/course/controller/v2/StudentCourseController.java:32]: removed.

File: api/src/main/java/ca/bc/gov/educ/api/course/controller/v2/StudentCourseController.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (cditcher) [reply api/src/main/java/ca/bc/gov/educ/api/course/controller/v2/StudentCourseController.java:73]: Need a separate endpoint for updates that uses PUT. We should not combine functionality on REST endpoints it breaks REST contract.

File: api/src/main/java/ca/bc/gov/educ/api/course/controller/v2/CourseController.java [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+        this.response = response;
+    }
+    @GetMapping(EducCourseApiConstants.GET_COURSE_BY_COURSE_ID_MAPPING)
+    @PreAuthorize(PermissionsConstants.READ_GRAD_COURSE)
+    @Operation(summary = "Find a Course by Course ID",
+            description = "Get a Course by Course ID", tags = { "Courses" })
+    @ApiResponses(value = {@ApiResponse(responseCode = "200", description = "OK"),
+            @ApiResponse(responseCode = "404", description = "NOT FOUND")})
+    public ResponseEntity<Course> getCourseDetails(@PathVariable String courseID) {
+        log.info("#getCourseDetails : courseID={}", courseID);
```
Comment (cditcher) [inline api/src/main/java/ca/bc/gov/educ/api/course/controller/v2/CourseController.java:49]: Change to debug level

File: api/src/main/java/ca/bc/gov/educ/api/course/controller/v2/CourseController.java [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+    }
+
+    @GetMapping(EducCourseApiConstants.GET_COURSE_BY_CODE_MAPPING)
+    @PreAuthorize(PermissionsConstants.READ_GRAD_COURSE)
+    @Operation(summary = "Find a Course by Course Code and Course Level",
+            description = "Get a Course by Course Code and Course Level", tags = { "Courses" })
+    @ApiResponses(value = {@ApiResponse(responseCode = "200", description = "OK"),
+            @ApiResponse(responseCode = "404", description = "NOT FOUND")})
+    public ResponseEntity<Course> getCourseDetails(@PathVariable String courseCode, @PathVariable String courseLevel) {
+        log.info("#getCourseDetails : courseCode={}, courseLevel={}", courseCode, courseLevel);
```
Comment (cditcher) [inline api/src/main/java/ca/bc/gov/educ/api/course/controller/v2/CourseController.java:60]: debug

File: api/src/main/java/ca/bc/gov/educ/api/course/controller/v2/StudentCourseController.java [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+        return response.UPDATED(result, StudentCourse.class);
+    }
+
+    @DeleteMapping(EducCourseApiConstants.STUDENT_COURSE_ID_MAPPING)
+    @PreAuthorize(PermissionsConstants.UPDATE_GRAD_COURSE_RESTRICTION)
+    @Operation(summary = "Delete a Student Course by Student Course ID", description = "Delete a Student Course by Student Course ID", tags = { "Student Courses" })
+    @ApiResponses(value = {@ApiResponse(responseCode = "204", description = "No Content"),
+            @ApiResponse(responseCode = "400", description = "BAD REQUEST"),
+            @ApiResponse(responseCode = "404", description = "NOT FOUND"),
+            @ApiResponse(responseCode = "422", description = "VALIDATION ERROR")
```
Comment (cditcher) [inline api/src/main/java/ca/bc/gov/educ/api/course/controller/v2/StudentCourseController.java:119]: Validation errors should be 400 as per https://www.rfc-editor.org/rfc/rfc7231#section-6.5.1

File: api/src/main/java/ca/bc/gov/educ/api/course/controller/v2/StudentCourseController.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (cditcher) [reply api/src/main/java/ca/bc/gov/educ/api/course/controller/v2/StudentCourseController.java:94]: Validation errors should be 400 as per https://www.rfc-editor.org/rfc/rfc7231#section-6.5.1

**Review by cditcher — CHANGES_REQUESTED** [review]

File: api/src/main/java/ca/bc/gov/educ/api/course/controller/v2/CourseController.java (hunk 8 above)
Comment (infstar) [reply api/src/main/java/ca/bc/gov/educ/api/course/controller/v2/CourseController.java]: done

File: api/src/main/java/ca/bc/gov/educ/api/course/controller/v2/StudentCourseController.java (hunk 10 above)
Comment (infstar) [reply api/src/main/java/ca/bc/gov/educ/api/course/controller/v2/StudentCourseController.java]: description is updated.

**Review by cditcher — APPROVED** [review]

**Pull request merged**